src/game/main_player.rs

```diff
@@ -1,23 +1,39 @@
-use crate::input::InputHandler;
+use crate::input::{InputHandler, InputObserver, InputCallback};
 use crate::render::camera::{Camera, PerspectiveCamera};
 
 use core::world::WorldCoordinate;
-use glutin::event::VirtualKeyCode;
 use math::vector::Vector3;
 
+mod input_handlers;
+
 const SPEED: f64 = 20.0;
 const SENSITIVITY: f32 = 0.2;
 
 pub struct MainPlayer {
     pub camera: PerspectiveCamera,
+    input_callbacks: Vec<InputCallback<Self>>,
+    is_moving_forward: bool,
+    is_moving_backward: bool,
+    is_moving_left: bool,
+    is_moving_right: bool,
+    is_moving_up: bool,
+    is_moving_down: bool,
```

Comment:
We should have something cleaner than this. This would not work with a joytick. Something something [GetAxis](https://docs.unity3d.com/ScriptReference/Input.GetAxis.html)